internal/venice-common/src/main/java/com/linkedin/venice/system/store/MetaStoreWriter.java

```diff
@@ -371,16 +373,52 @@ private void update(
     }
     StoreMetaKey key = dataType.getStoreMetaKey(keyStringSupplier.get());
     SpecificRecord update = updateSupplier.get();
-    writer.update(
-        key,
-        update,
-        AvroProtocolDefinition.METADATA_SYSTEM_SCHEMA_STORE.currentProtocolVersion.get(),
-        derivedComputeSchemaId,
-        null);
-    writer.flush();
+    writeMessageWithRetry(metaStoreName, vw -> {
+      vw.update(
+          key,
+          update,
+          AvroProtocolDefinition.METADATA_SYSTEM_SCHEMA_STORE.currentProtocolVersion.get(),
+          derivedComputeSchemaId,
+          null);
+    });
+  }
+
+  void writeMessageWithRetry(String metaStoreName, Consumer<VeniceWriter> writerConsumer) {
+    ReentrantLock lock = getOrCreateMetaStoreWriterLock(metaStoreName);
+    int messageProduceRetryCount = 0;
+    int maxMessageProduceRetryCount = 3;
+    VeniceWriter writer = null;
+    boolean messageProduced = false;
+    while (!messageProduced) {
+      try {
+        lock.lock();
+        writer = getOrCreateMetaStoreWriter(metaStoreName);
+        writerConsumer.accept(writer);
+        writer.flush();
+        messageProduced = true;
+      } catch (Exception e) {
+        messageProduceRetryCount++;
+        if (messageProduceRetryCount < maxMessageProduceRetryCount) {
+          LOGGER.warn(
+              "Caught exception while trying to write message, will restart Venice Writer and retry {}/{}",
+              messageProduceRetryCount,
+              maxMessageProduceRetryCount);
+          closeVeniceWriter(metaStoreName, writer, true);
```

Comment:
Is it possible that `closeVeniceWriter` could throw another exception?
Can we catch the potential exception to avoid breaking the loop?

Reply:
Hi Gaojie, I don't think there is any exception to be thrown here, the closeVeniceWrtier with skipTopicCheck = true means it will just call vw.close() with graceful close = true.
Inside VW.close() it exception is caught and handled with WARN logging.

Reply:
This is mainly for safety purpose.
There is no guarantee that function won't throw exception in the future, right?